lib/semantics/check-do.cc

```diff
@@ -833,6 +834,50 @@ void DoChecker::Leave(const parser::ConnectSpec &connectSpec) {
   }
 }
 
+using ActualArgumentSet = std::set<evaluate::ActualArgumentRef>;
+
+struct CollectActualArgumentsHelper
+  : public evaluate::SetTraverse<CollectActualArgumentsHelper,
+        ActualArgumentSet> {
+  using Base = SetTraverse<CollectActualArgumentsHelper, ActualArgumentSet>;
+  CollectActualArgumentsHelper() : Base{*this} {}
+  using Base::operator();
+  ActualArgumentSet operator()(const evaluate::ActualArgument &arg) const {
+    return ActualArgumentSet{arg};
+  }
+};
+
+template<typename A> ActualArgumentSet CollectActualArguments(const A &x) {
+  return CollectActualArgumentsHelper{}(x);
+}
+
+template ActualArgumentSet CollectActualArguments(const SomeExpr &);
+
+void DoChecker::Leave(const parser::Expr &parserExpr) {
+  if (parserExpr.typedExpr) {
+    if (const SomeExpr * expr{GetExpr(parserExpr)}) {
+      ActualArgumentSet argSet{CollectActualArguments(*expr)};
+      for (const evaluate::ActualArgumentRef &argRef : argSet) {
+        if (const SomeExpr * argExpr{argRef->UnwrapExpr()}) {
+          if (evaluate::IsVariable(*argExpr)) {
+            if (const Symbol * symbol{evaluate::GetFirstSymbol(*argExpr)}) {
+              switch (argRef->dummyIntent()) {
```

Comment:
Would this be better as an if statement?

Since argRef is available before the three preceding if statements can we consider hoisting this up or putting a continue for the negation (INTENT(IN) or default intent)?

Reply:
Good idea.  That'll make the code a little faster for most cases.